pkg/kv/dynamodb/store.go

```diff
@@ -0,0 +1,320 @@
+		ItemKey:   string(key),
+		ItemValue: string(value),
+	}
+
+	marshaledItem, err := dynamodbattribute.MarshalMap(item)
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+
+	input := &dynamodb.PutItemInput{
+		Item:      marshaledItem,
+		TableName: &s.params.TableName,
+	}
+	if usePredicate {
+		if valuePredicate != nil {
+			input.ConditionExpression = aws.String(ItemValue + " = :predicate")
+			input.ExpressionAttributeValues = map[string]*dynamodb.AttributeValue{
+				":predicate": {S: aws.String(string(valuePredicate.([]byte)))},
+			}
+		} else {
+			input.ConditionExpression = aws.String("attribute_not_exists(" + ItemValue + ")")
+		}
+	}
+
+	_, err = s.svc.PutItemWithContext(ctx, input)
+
+	if err != nil {
+		if _, ok := err.(*dynamodb.ConditionalCheckFailedException); ok && usePredicate {
+			return kv.ErrPredicateFailed
+		}
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+	return nil
+}
+
+func (s *Store) Delete(ctx context.Context, key []byte) error {
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+
+	_, err := s.svc.DeleteItemWithContext(ctx, &dynamodb.DeleteItemInput{
+		TableName: aws.String(s.params.TableName),
+		Key:       s.bytesKeyToDynamoKey(key),
+	})
+
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+	return nil
+}
+
+func (s *Store) Scan(ctx context.Context, start []byte) (kv.EntriesIterator, error) {
+	return s.scanInternal(ctx, start, nil)
+}
+
+func (s *Store) scanInternal(ctx context.Context, scanKey []byte, exclusiveStartKey map[string]*dynamodb.AttributeValue) (*EntriesIterator, error) {
+	keyConditionExpression := PartKey + " = :partkey"
+	expressionAttributeValues := map[string]*dynamodb.AttributeValue{
+		":partkey": {
+			S: aws.String(s.getPartitionKey(scanKey)),
+		},
+	}
+	if len(scanKey) > 0 {
+		keyConditionExpression += " AND " + ItemKey + " >= :fromkey"
+		expressionAttributeValues[":fromkey"] = &dynamodb.AttributeValue{
+			S: aws.String(string(scanKey)),
+		}
+	}
+	queryInput := &dynamodb.QueryInput{
+		TableName:                 aws.String(s.params.TableName),
+		KeyConditionExpression:    aws.String(keyConditionExpression),
+		ExpressionAttributeValues: expressionAttributeValues,
+		ConsistentRead:            aws.Bool(true),
+		ScanIndexForward:          aws.Bool(true),
+		ExclusiveStartKey:         exclusiveStartKey,
+	}
+	queryOutput, err := s.svc.QueryWithContext(ctx, queryInput)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w (start=%v)", err, kv.ErrOperationFailed, string(scanKey))
+	}
+
+	return &EntriesIterator{
+		store:        s,
+		partKey:      s.getPartitionKey(scanKey),
+		startKey:     string(scanKey),
+		queryResult:  queryOutput,
+		currEntryIdx: 0,
+		err:          nil,
+	}, nil
+}
+
+func (s *Store) Close() {}
+
+func (e *EntriesIterator) Next() bool {
+	if e.err != nil {
+		return false
+	}
+
+	if e.currEntryIdx == int(*e.queryResult.Count) {
+		if e.queryResult.LastEvaluatedKey == nil {
```

Comment:
how can this happen?

Reply:
`Query` (`QueryWithContext`) will return `Count` items. If there are more items that satisfy the `KeyConditionExpression` the `LastEvaluatedKey` is used as a `ContinuationToken` which can be used for paging.
With each "page", I'm using the `currEntryIdx` to loop over the objects in that page. When `currEntryIdx` equals `Count` the end of the page is reached. If this is the last page the above condition will apply, signifying the last item for that partition is reached